File: js/cloud.ts
import rl from "raylib"

export type Cloud = {
    x: number
    y: number
    signature: number[]
}

export const cloudBaseSizeX = 160
export const cloudBaseSizeY = 160

export const importedCloudBaseSizeX = 300
export const importedCloudBaseSizeY = 300

export function createCloud(x: number, y: number): Cloud {
    return {
        x,
        y,
        signature: generateSignature(),
    }
}

export function createCloudWithSignature(x: number, y: number, hash: string): Cloud {
    return {
        x,
        y,
        signature: parseSignature(hash),
    }
}

export function drawCloud(cloud: Cloud) {
    for (let i = 0; i < cloud.signature[0]; i++) {
        const size = cloud.signature[i * 3 + 1]
        const left = cloud.signature[i * 3 + 2]
        const top = cloud.signature[i * 3 + 3]
        
        rl.DrawCircle(cloud.x + left + size / 2, cloud.y + top + size / 2, size, rl.WHITE)
    }
}

function parseSignature(hash: string): number[] {
    return hash.split(":").map(x => parseInt(x))
}

function generateSignature(): number[] {
    const res: number[] = []
	
	const numCircle = Math.random() * 10 + 10
    res.push(numCircle)
	
	for (let a = 0; a < numCircle; a++) {
		const size = Math.random() * 50 + 15
		const left = Math.random() * cloudBaseSizeX
		const top = Math.random() * cloudBaseSizeY
		
        res.push(size)
        res.push(left)
        res.push(top)
	}
	
	return res
}

File: mlang/cloud.m
import "lib/raylib.m"

pub struct Cloud {
    x: f32
    y: f32
    signature: [2]i32
}

const baseSizeX = 160
const baseSizeY = 160

pub fn createCloud(x: f32, y: f32): Cloud {
    return .{
        x = x
        y = y
        signature = generateSignature()
    }
}

pub fn drawCloud(cloud: &Cloud) {
    for cloud.signature[0] |i| {
        const size = cloud.signature[i * 3 + 1]
        const left = cloud.signature[i * 3 + 2]
        const top = cloud.signature[i * 3 + 3]
        
        drawCircle(<i32>(cloud.x + left), <i32>(cloud.y + top), <f32>size, WHITE)
    }
}

fn generateSignature(): [dyn]i32 {
    val res: [dyn]i32 = .[]
    
    const numCircle = random(0, 10) + 10
    res.add(numCircle)
    
    for numCircle {
        const size = random(0, 50) + 15
        const left = random(0, baseSizeX)
        const top = random(0, baseSizeY)
        
        res.add(size)
        res.add(left)
        res.add(top)
    }
    
    return res
}